### ISSUE TITLE
Results not on a regular time grid

### ISSUE THREAD

**Issue by alexa-f** [CATIA-Systems/FMPy#28]
Hi,

I need the results to reside on a regular grid and I was expecting this behaviour since I explicitly specified the `output_interval`. Observing the actual output I found three issues, where I am not sure if this is intended by the author:

1. doesn't start at t=0s,
2. some samples are droped, e.g. at t=0.06s,
3. grid is shifted starting after t=0.42s.

Could you please clarify? Thanks.

# Simple example:
```python
from fmpy import simulate_fmu

filename = 'fmus/Dymola2017/CoupledClutches.fmu'

print(filename)
results = simulate_fmu(
    filename=filename,
    start_time=0,
    stop_time=1.5,
    solver='CVode',
    step_size=1e-2,
    output_interval=2e-2,
    input=None,
    output=None,
    validate=False,
    fmi_logging=False)

print(results['time'])
```

# Output:
```python
[ 0.02        0.04        0.08        0.1         0.12        0.14        0.18
  0.22        0.26        0.28        0.3         0.32        0.34        0.36
  0.38        0.4         0.42        0.44565285  0.46565285  0.49740768
  0.53740768  0.55740768  0.57740768  0.59740768  0.63429745  0.65429745
  0.67429745  0.70976556  0.72976556  0.74976556  0.76976556  0.78976556
  0.82000019  0.85116724  0.87116724  0.89619621  0.92065559  0.94296037
  0.96296037  0.98296037  1.00355202  1.03929382  1.05929382  1.07929382
  1.11029498  1.13029498  1.15733956  1.17733956  1.21676477  1.23676477
  1.27071325  1.29071325  1.31071325  1.33071325  1.3697638   1.3897638
  1.4097638   1.4297638   1.4497638   1.4701323   1.4901323   1.5       ]
```

**Comment by t-sommer** [comment]
The `CVode` solver influences the actual output interval. If you need the sample points to be regular you can either use a co-simulation FMU or the `Euler` solver. Note that `step_size` is only used for Euler (see [docstring](https://github.com/CATIA-Systems/FMPy/blob/master/fmpy/simulation.py#L336)).

**Comment by alexa-f** [comment]
In `simulateME()` within `simulation.py` I see that the solver returns a new time value, which is recorded some lines later. I am with you, this is how the solver works.

```python
time = solver.step(time, tNext)

# record the values for this step
recorder.sample(time)
```

But what about the expected behaviour as implied by the description `interval for sampling the output` for `output_interval` (see [docstring](https://github.com/CATIA-Systems/FMPy/blob/master/fmpy/simulation.py#L336)). Would it be possible to do some additional interpolation step after simulation is finished, in order to output the results on a regular grid?

**Comment by t-sommer** [comment]
Why not just use a co-simulation FMU? It will skip any intermediate samples (e.g. at events) and interpolate the values at the (regular) sample points.

**Comment by chrbertsch** [comment]
Wouldn't it be good to include the intial time point (t=0) in the output vector?
Dymola e.g. expects and uses a commucation step size, which is used additionally to event times to display. The velues are generated via interpolation from the solver timesteps. Advantage: significantly reduced output data size.
Difference to a CS FMU: at each communication timestep the variable step-solver integration is not interrupted  (so there must not be small timestep around the communication point)

**Comment by t-sommer** [comment]
> Wouldn't it be good to include the intial time point (t=0) in the output vector?

Yes. I've opened a new issue (#29)

**Comment by t-sommer** [comment]
> Difference to a CS FMU: at each communication timestep the variable step-solver integration is not interrupted (so there must not be small timestep around the communication point)

This is already the case in [FMPy (/w CVode)](https://github.com/CATIA-Systems/FMPy/blob/master/fmpy/sundials/__init__.py#L249) and Dymola co-simulation FMUs.

```python
flag = CVode(self.cvode_mem, tNext, self.x, byref(tret), CV_NORMAL)
```

From the [cv_guide.pdf](https://computation.llnl.gov/sites/default/files/public/cv_guide.pdf), 4.5.5:

[...] The CV_NORMAL option causes the solver to take internal steps until it has reached
or just passed the user-specified tout parameter. The solver then interpolates
in order to return an approximate value of y(tout). The CV_ONE_STEP
option tells the solver to take just one internal step and then return the
solution at the point reached by that step.

**Comment by chrbertsch** [comment]
I would not regard this as the expected behaviour for a CS FMU. 
If the "communcation times" correspond to the "doSteps", I would not expect an interpolated value, but the correct value. So there should be really a macro-timestep corresponding to the doStep.
In a co-simulation context, one could get new inputs at the communication points, so that this point must be hit.
FMI for CS is not intended for asynchronous co-simulation, but there must be a synchronization at each macro-timestep.

**Comment by t-sommer** [comment]
I've just pushed my [first attempt](a85c6fae) to implement your suggestions. It required a lot of changes to the ME infrastructure so I'd like to invite you to test and review the changes before merging them.

**Comment by alexa-f** [comment]
I checked the new test `test_output_grid` and it gives the expected results. Thanks a lot for this work!
The only thing I noticed is a bug, if no input is specified:
```python
Traceback (most recent call last):
  File "C:/Users/xxx/Documents/git_projects/FMPy/test_output_grid.py", line 27, in <module>
    result = simulate_fmu(solver='Euler', record_events=False, **kwargs)
  File "C:\Users\xxx\Documents\git_projects\FMPy\fmpy\simulation.py", line 450, in simulate_fmu
    result = simulateME(model_description, fmu_args, start_time, stop_time, solver, step_size, relative_tolerance, start_values, input, output, output_interval, record_events, timeout, callbacks, step_finished)
  File "C:\Users\xxx\Documents\git_projects\FMPy\fmpy\simulation.py", line 594, in simulateME
    if t_next - time > eps:
TypeError: unsupported operand type(s) for -: 'NoneType' and 'float'
```

In this case `t_input_event` is `None`.
```pyhton
t_input_event = input.apply(time)
```

I fixed it by 
```pyhton
if t_input_event:
    input_event = t_input_event <= t_next
else:
    input_event = None
```

**Comment by t-sommer** [comment]
I've merged the feature into develop with 0fd04161eb391fcffc8a6ce17cf1a05350debac6 that also fixes for the problem you've found. Thanks for testing!
